### PR TITLE
Leagues signals and models enhancements

### PR DESCRIPTION
**Pull request by mazdakb**
### Change Log:
* Enhanced file upload and overwrite behaviors for duplicate handling
* Add MEDIA_URL and serving in DEBUG mode
* Add MEDIA_ROOT to .gitignore
* Separate signals from models
* Add OverwriteStorage backend
* Add test cases for files delete and overwrite
* Add test libraries to requirements file
* Pinned all library versions in requirements file
* Add django-environ
* Read database config from .env
* Add .env to .gitignore

```
## Sample database config for .env

# PostgreSQL
POSTGRES_PASSWORD=ares
POSTGRES_USER=ares
```

### CODE REVIEW

**Review by mehdy — COMMENTED** [review]
Thanks!
these are pretty good points. I'll merge it as soon as you resolve the review I wrote for you.

p.s for your next contributions please keep your PRs small and send more :) it's easier to review, merge and etc.

File: ares/urls.py [hunk 1]
```diff
@@ -3,8 +3,10 @@
...
 from django.conf.urls import url, include
 from django.contrib import admin
+from django.conf import settings
+from django.conf.urls.static import static
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
     url('^accounts/', include('accounts.urls')),
-]
+] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
Comment (geraneum) [inline ares/urls.py]: I understand your security standpoint but I put it here for following reasons. If you're not convinced at the end I can add `static()` in DEBUG mode like this:
```
if settings.DEBUG:
    urlpatterns += static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)
```
However my reasons are:
1. We want to be able to serve other media files (any other user uploaded content like their profile picture maybe).
2. Serving `code` files to their owners also should be ok. (right?)
3. We need file serving functionality in development environment without a reverse proxy.
4. We will be using a reverse proxy in production in front of Django (i.e Nginx)

With these in mind we can use one of these two solutions:
* Use X-Sendfile (X-Accel on Nginx) to handle file serving authorization efficiently. For reference, please take a look at [Official Nginx Docs](https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/) or [this post](https://wellfire.co/learn/nginx-django-x-accel-redirects/) or [django-sendfile](https://github.com/johnsensible/django-sendfile)
* Save `code` files to a location other than MEDIA_ROOT (by doing this serving them to their owners needs extra care)

P.S I suggest the former

**Comment by geraneum** [comment]
@mehdy Sure, I'll try to break my PRs down next time.

File: ares/urls.py (hunk 1 above)
Comment (mehdy) [inline ares/urls.py]: you're right. it's ok for now. I'll merge this now but I'm also opening a security issue about this that must be resolved before the first deployment.

and I like the nginx way. but It doesn't really differentiate.

**Review by mehdy — APPROVED** [review]
everything's fine. this issue #33 is generated here

**Pull request merged**